packages/firebase_database/lib/src/database_reference.dart

```diff
@@ -125,10 +125,46 @@ class DatabaseReference extends Query {
   ///
   /// remove() is equivalent to calling set(null)
   Future<Null> remove() => set(null);
+
+  /// Determine the next transaction key to be used. The length of the
+  /// transactions map is used to ensure uniqueness of keys.
+  int getNextTransactionKey() {
+    return FirebaseDatabase._transactions.length;
+  }
+
+  /// Performs an optimistic-concurrency transactional update to the data at
+  /// this Firebase Database location.
+  Future<Null> runTransaction(TransactionHandler transactionHandler) {
+    final int transactionKey = getNextTransactionKey();
+    FirebaseDatabase._transactions[transactionKey] = transactionHandler;
+    return _database._channel.invokeMethod(
+        'DatabaseReference#runTransaction', <String, dynamic>{
+      'path': path,
+      'transactionKey': transactionKey
+    }).then((Map<String, dynamic> result) {
+      final DataSnapshot dataSnapshot = new DataSnapshot._(result);
+      return transactionHandler.doTransaction(dataSnapshot);
```

Comment:
Since a `Future` can only complete once, I don't think this handles the case where the first attempt at `doTransaction` fails (because data was changed on the server) and we end up getting called with a different snapshot.

Reply:
Updated runTransaction to handle the case of multiple calls of doTransaction.